internal/btf/btf.go

```diff
@@ -203,42 +203,35 @@ func loadKernelSpec() (*Spec, error) {
 		return nil, fmt.Errorf("can't read kernel release number: %w", err)
 	}
 
+	fh, err := os.Open("/sys/kernel/btf/vmlinux")
+	if err == nil {
+		defer fh.Close()
+
+		return loadNakedSpec(fh, internal.NativeEndian, nil, nil)
+	}
+
 	// use same list of locations as libbpf
 	// https://github.com/libbpf/libbpf/blob/9a3a42608dbe3731256a5682a125ac1e23bced8f/src/btf.c#L3114-L3122
-	locations := []struct {
-		pathFmt string
-		raw     bool
-	}{
-		/* try canonical vmlinux BTF through sysfs first */
-		{"/sys/kernel/btf/vmlinux", true /* raw BTF */},
-		/* fall back to trying to find vmlinux ELF on disk otherwise */
-		{"/boot/vmlinux-%s", false},
-		{"/lib/modules/%s/vmlinux-%s", false},
-		{"/lib/modules/%s/build/vmlinux", false},
-		{"/usr/lib/modules/%s/kernel/vmlinux", false},
-		{"/usr/lib/debug/boot/vmlinux-%s", false},
-		{"/usr/lib/debug/boot/vmlinux-%s.debug", false},
-		{"/usr/lib/debug/lib/modules/%s/vmlinux", false},
-	}
-
-	var spec *Spec
+	locations := []string{
+		"/boot/vmlinux-%s",
+		"/lib/modules/%s/vmlinux-%[1]s",
+		"/lib/modules/%s/build/vmlinux",
+		"/usr/lib/modules/%s/kernel/vmlinux",
+		"/usr/lib/debug/boot/vmlinux-%s",
+		"/usr/lib/debug/boot/vmlinux-%s.debug",
+		"/usr/lib/debug/lib/modules/%s/vmlinux",
+	}
+
 	for _, loc := range locations {
-		path := fmt.Sprintf(loc.pathFmt, release)
+		path := fmt.Sprintf(loc, release)
+
 		fh, err := os.Open(path)
 		if err != nil {
 			continue
 		}
 		defer fh.Close()
-		if loc.raw {
-			spec, err = loadNakedSpec(fh, internal.NativeEndian, nil, nil)
-		} else {
-			spec, err = loadNakedSpecFromReader(fh, nil, nil)
-		}
 
-		if err != nil {
-			continue
-		}
-		return spec, nil
+		return loadSpecFromVmlinux(fh)
```

Comment:
There isn't a guarantee that these files will have a `.BTF` section. Should we keep searching if that is the case?

Reply:
I think that's unlikely to help: most systems will only have a single path where vmlinux is located (since it's a single package) so trying other location is unlikely to help. Trying other paths that also has the downside that we return ErrNotFound instead of an error describing why we can't use the existing BTF. That is really confusing to users (and to me when I started to look into the issue).